File: leetcode/algorithms/405_convert_a_number_to_hexadecimal/main.c
#include <stdlib.h>

char* toHex(int num) {
    char* hex = "0123456789abcdef";
    unsigned int unum = num;
    int count = 0;
    char* result = (char*)malloc(sizeof(char) * 9);
    result[8] = '\0';
    while (1) {
        result[7 - count] = hex[unum % 16];
        unum >>= 4;
        count++;
        if (!unum) {
            break;
        }
    }
    return &result[8 - count];
}


// Solution
char* solution(int num) {
    char* hex = (char*)malloc(sizeof(char) * 10);
    char nib;
    int i = 0;
    unsigned int x = num;
    hex[9] = '\0';
    while (1) {
        nib = (x & 0xF);
        hex[8 - i] = (nib > 9) ? ('a' + nib - 10) : ('0' + nib);
        x >>= 4;
        if (!x) break;
        i++;
    }
    return &hex[8 - i];
}

File: leetcode/algorithms/405_convert_a_number_to_hexadecimal/main.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class ConvertANumberToHexadecimal {
    public String toHex(int num) {
        long longNum = num;
        List<String> hex = new ArrayList<String>(
                Arrays.asList("0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "a", "b", "c", "d", "e", "f"));

        if (longNum == 0) {
            return hex.get(0);
        }

        if (longNum < 0) {
            longNum += Math.pow(2, 32);
        }

        String result = "";

        while (longNum > 0) {
            long remainder = longNum % 16;
            result = hex.get((int) remainder) + result;
            longNum = Math.floorDiv(longNum, 16);
        }

        return result;
    }


    // Solution
    public String solution(int num) {
        char[] map = { '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f' };

        if (num == 0) {
            return "0";
        }

        String result = "";
        while (num != 0) {
            result = map[(num & 15)] + result;
            num = (num >>> 4);
        }

        return result;
    }
}